driver-core/src/main/com/mongodb/internal/operation/DropCollectionOperation.java

```diff
@@ -85,52 +97,183 @@ public WriteConcern getWriteConcern() {
+        SingleResultCallback<Void> errHandlingCallback = errorHandlingCallback(callback, LOGGER);
+        getEncryptedFields((AsyncReadWriteBinding) binding, (result, t) -> {
+            if (t != null) {
+                errHandlingCallback.onResult(null, t);
+            } else {
+                withAsyncConnection(binding, (connection, t1) -> {
+                    if (t1 != null) {
+                        errHandlingCallback.onResult(null, t1);
+                    } else {
+                        new ProcessCommandsCallback(binding, connection, getCommands(result), releasingCallback(errHandlingCallback,
+                                connection))
+                                .onResult(null, null);
+                    }
+                });
             }
         });
     }
 
-    private BsonDocument getCommand(final ConnectionDescription description) {
+    /**
+     * With FLE2 dropping a collection can involve more logic and commands.
+     *
+     * <p>
+     * A call to a driver helper Collection.drop(dropOptions) must check if the collection namespace (<databaseName>.<collectionName>)
+     * has an associated encryptedFields. Check for an associated encryptedFields from the following:
+     * <ol>
+     *     <li>The encryptedFields option passed in dropOptions.</li>
+     *     <li>The value of AutoEncryptionOpts.encryptedFieldsMap[<databaseName>.<collectionName>].</li>
+     *     <li>If AutoEncryptionOpts.encryptedFieldsMap is not null, run a listCollections command on the database databaseName with the
+     *          filter { "name": "<collectionName>" }. Check the returned options for the encryptedFields option.</li>
+     * </ol>
+     * </p>
+     * <p>
+     * If the collection namespace has an associated encryptedFields, then do the following operations.
+     * If any of the following operations error, the remaining operations are not attempted:
+     * <ol>
+     * <li>Drop the collection collectionName.
+     * <li>Drop the collection with name encryptedFields["escCollection"].
+     *    If encryptedFields["escCollection"] is not set, use the collection name enxcol_.<collectionName>.esc.</li>
+     * <li>Drop the collection with name encryptedFields["eccCollection"].
+     *    If encryptedFields["eccCollection"] is not set, use the collection name enxcol_.<collectionName>.ecc.</li>
+     * <li>Drop the collection with name encryptedFields["ecocCollection"].
+     *    If encryptedFields["ecocCollection"] is not set, use the collection name enxcol_.<collectionName>.ecoc.</li>
+     * </ol>
+     * </p>
+     *
+     * @return the list of commands to run to create the collection
+     */
+    private List<Function<ConnectionDescription, BsonDocument>> getCommands(final BsonDocument encryptedFields) {
+        if (encryptedFields == null) {
+            return singletonList(this::dropCollectionCommand);
+        } else  {
+            return asList(
+                    this::dropCollectionCommand,
```

Comment:
Per https://github.com/mongodb/specifications/pull/1215/, do the drop of the user collection last. We warn users in the log file to clean up the esc/ecc/ecoc collections if they drop the user collection first.

Reply:
Ack will do in JAVA-4609 - currently blocked until merge.